File: Java/Java/src/SRT.java
public class SRT {

	static void findWaitingTime(int[] processes, int[] bt, int[] wt, int quantum) {
		int n = processes.length;
		int[] rem_bt = bt.clone();
		int t = 0;
		while (true) {
			boolean done = true;
			for (int i = 0; i < n; i++) {
				if (rem_bt[i] > 0) {
					done = false;
					if (rem_bt[i] > quantum) {
						t += quantum;
						rem_bt[i] -= quantum;
					} else {
						t += rem_bt[i];
						wt[i] = t - bt[i];
						rem_bt[i] = 0;
					}
				}
			}
			if (done) break;
		}
	}

	static void findTurnAroundTime(int[] processes, int[] bt, int[] wt, int[] tat) {
		int n = processes.length;
		for (int i = 0; i < n; i++)
			tat[i] = bt[i] + wt[i];
	}

	static void srt(int[] processes, int[] bt, int quantum) {
		int n = processes.length;
		int[] wt = new int[n], tat = new int[n];
		long startTime = System.nanoTime();
		findWaitingTime(processes, bt, wt, quantum);
		findTurnAroundTime(processes, bt, wt, tat);
		long endTime = System.nanoTime();
		double total_wt = 0, total_tat = 0;
		for (int i = 0; i < n; i++) {
			total_wt += wt[i];
			total_tat += tat[i];
		}
		double avg_wt = total_wt / n;
		double avg_tat = total_tat / n;
		double throughput = n / ((endTime - startTime) / 1_000_000_000.0);
		System.out.println("Processes Burst Time Waiting Time Turn-Around Time");
		for (int i = 0; i < n; i++)
			System.out.printf(" %d         %d        %d          %d%n", processes[i], bt[i], wt[i], tat[i]);
		System.out.printf("%nAverage waiting time = %.5f%n", avg_wt);
		System.out.printf("Average turn around time = %.5f%n", avg_tat);
		System.out.printf("Throughput = %.5f%n", throughput);
		System.out.printf("Running time = %.5f seconds%n", (endTime - startTime) / 1_000_000_000.0);
	}

	public static void main(String[] args) {
		int[] processes = {1, 2, 3};
		int[] burst_time = {10, 5, 8};
		int quantum = 2;
		srt(processes, burst_time, quantum);
	}
}

File: Python/SRT.py
import time

def findWaitingTime(processes, burst_times, arrival_times, wt, quantum):
	n = len(processes)
	rt = burst_times.copy()
	complete = 0
	t = 0
	minm = float('inf')
	short = 0
	check = False

	while complete != n:
		for j in range(n):
			if arrival_times[j] <= t and rt[j] < minm and rt[j] > 0:
				minm = rt[j]
				short = j
				check = True
		if not check:
			t += 1
			continue

		rt[short] -= 1
		minm = rt[short] if rt[short] > 0 else float('inf')

		if rt[short] == 0:
			complete += 1
			check = False
			fint = t + 1
			wt[short] = fint - burst_times[short] - arrival_times[short]
			if wt[short] < 0:
				wt[short] = 0
		t += 1

	return t  # Total running time

def findTurnAroundTime(burst_times, wt, tat):
	n = len(burst_times)
	for i in range(n):
		tat[i] = burst_times[i] + wt[i]

def findavgTime(processes, burst_times, arrival_times, quantum):
	n = len(processes)
	wt = [0] * n
	tat = [0] * n

	start_time = time.time()
	total_running_time = findWaitingTime(processes, burst_times, arrival_times, wt, quantum)
	findTurnAroundTime(burst_times, wt, tat)
	end_time = time.time()

	print("Processes    Burst Time     Waiting Time    Turn-Around Time")
	total_wt = 0
	total_tat = 0
	for i in range(n):
		total_wt += wt[i]
		total_tat += tat[i]
		print(f" {processes[i]}           {burst_times[i]}            {wt[i]}            {tat[i]}")

	avg_wt = total_wt / n
	avg_tat = total_tat / n
	throughput = n / total_running_time
	running_duration = end_time - start_time

	print("\nAverage waiting time = %.5f" % avg_wt)
	print("Average turn around time = %.5f" % avg_tat)
	print("Total running time (CPU clock time) = %.5f seconds" % running_duration)
	print("Throughput = %.5f processes per time unit" % throughput)

if __name__ == "__main__":
	processes = [1, 2, 3, 4]
	burst_times = [6, 8, 7, 3]
	arrival_times = [1, 1, 2, 3]
	quantum = 2
	findavgTime(processes, burst_times, arrival_times, quantum)
